### PR TITLE
Fix selections that have inner inputs.

### PR DESCRIPTION
**Pull request by llllllllll**
Things like `a[a.a == b.a]` would fail because it wouldn't compute the `b.a.`

### CODE REVIEW

File: blaze/compute/sql.py [hunk 1]
```diff
@@ -262,19 +262,41 @@ def compute_up(t, s, **kwargs):
 @dispatch(Selection, sa.sql.ColumnElement)
 def compute_up(expr, data, scope=None, **kwargs):
     predicate = compute(expr.predicate, data, post_compute=False)
-    return sa.select([data]).where(predicate)
+    return compute_up(expr, data, predicate, scope=scope, **kwargs)
+
+
+@dispatch(Selection, sa.sql.ColumnElement, ClauseElement)
```
Comment (cpcloud) [inline blaze/compute/sql.py]: Should `ClauseElement` be `sa.sql.expression.BinaryExpression`?

Comment (cpcloud) [inline blaze/compute/sql.py]: `ClauseElement` is very general and would let basically any sqlalchemy expression in.

File: blaze/expr/broadcast.py [hunk 2]
```diff
@@ -125,15 +127,20 @@ def broadcast_collect(expr, Broadcastable=Broadcastable,
...
+
     # Recurse down
-    children = [broadcast_collect(i, Broadcastable, WantToBroadcast)
-                for i in expr._inputs]
-    return expr._subs(dict(zip(expr._inputs, children)))
+    children = (
+        broadcast_collect(i, broadcastable, want_to_broadcast, no_recurse)
+        for i in expr._inputs
+    )
+    return expr._subs({e: c for e, c in zip(expr._inputs, children)})
```
Comment (cpcloud) [inline blaze/expr/broadcast.py]: You couldn't resist :D

File: blaze/compute/sql.py (hunk 1 above)
Comment (llllllllll) [inline blaze/compute/sql.py]: Is `BinaryExpression` too restrictive here? Are there any valid predicate functions that don't fall into this? Is there any `ClauseElement` that we would not treat as a valid predicate?

File: blaze/expr/broadcast.py (hunk 2 above)
Comment (llllllllll) [inline blaze/expr/broadcast.py]: yeah, I saw my chance to take advantage of dropping 2.6 and I went for it.

**Comment by llllllllll** [comment]
looks like I broke everything. I will clean the other backends up and ping you when its passing.

**Comment by llllllllll** [comment]
related #1230

I need to get back to this. I haven't had much time to work on blaze recently ;_;

**Comment by llllllllll** [comment]
Hoping the tests pass. This should be ready to review so I am removing the wip tag.

**Comment by llllllllll** [comment]
@cpcloud any thoughts on this? Currently using this for our deploy so it would be good to get this merged.

**Comment by cpcloud** [comment]
I'll review this today n

File: blaze/compute/sql.py (hunk 1 above)
Comment (cpcloud) [inline blaze/compute/sql.py]: Yes, `Selectable` is a subclass of `ClauseElement` and we don't support non-column boolean expressions.

Comment (llllllllll) [inline blaze/compute/sql.py]: I tried making this `BinaryExpression` but it looks like there are some cases where we are getting a `BooleanClauseList`. The first common parrent is `ColumnElement`, does that seem restrictive enough?

Comment (cpcloud) [inline blaze/compute/sql.py]: Yeah, `ColumnElement` is fine

**Comment by llllllllll** [comment]
Not sure if the tests are still failing with a 500 from binstar, any idea what caused (or is still causing) this?

**Comment by cpcloud** [comment]
+1 here

**Comment by llllllllll** [comment]
> +1

good to merge?

**Pull request merged**